wai-extra/Network/Wai/Parse.hs

```diff
@@ -796,17 +796,35 @@ sinkTillBound bound iter seed0 src max' = do
     b <- final
     return (b, seed)
 
-parseAttrs :: S.ByteString -> [(S.ByteString, S.ByteString)]
-parseAttrs = map go . S.split _semicolon
-  where
-    tw = S.dropWhile (== _space)
-    dq s =
-        if S.length s > 2 && S.head s == _quotedbl && S.last s == _quotedbl
-            then S.tail $ S.init s
-            else s
-    go s =
-        let (x, y) = breakDiscard _equal s
-         in (tw x, dq $ tw y)
+parseContentDispositionAttrs :: S.ByteString -> [(S.ByteString, S.ByteString)]
+parseContentDispositionAttrs = parseTokenValues
+ where
+    nonTokenChars = [_semicolon, _equal]
+    dropSpace = S.dropWhile (== _space)
+    parseTokenValues input | S.null input = []
+    parseTokenValues input =
+        let (token, rest) = parseToken $ dropSpace input
+         in case S.uncons rest of
+            Just (c, rest')
+                | c == _equal -> 
+                    let (value, rest'') = parseValue $ dropSpace rest'
+                     in (token, value) : parseTokenValues (S.drop 1 rest'')
+                | otherwise -> (token, S.empty) : parseTokenValues rest'
+            Nothing -> (token, S.empty) : parseTokenValues S.empty
+    parseToken = S.break (`elem` nonTokenChars)
+    parseValue input =
+        case S.uncons $ dropSpace input of
+            Just (c, rest) | c == _quotedbl -> parseQuotedString [] rest
+            _ -> S.break (`elem` nonTokenChars) $ dropSpace input
+    parseQuotedString acc input =
+        let (prefix, rest) = S.break (`elem` [_quotedbl, _backslash]) input
+         in case S.uncons rest of
+            Just (c, rest')
+                | c == _quotedbl -> (S.concat $ reverse (prefix:acc), rest')
+                | c == _backslash ->
+                    let (slashed, postSlash) = S.splitAt 1 rest'
+                     in parseQuotedString (slashed:prefix:acc) postSlash
+            _ -> (prefix, rest)
```

Comment:
This doesn't use the acc, so there's potential for loss of information.

I think you'll lose information if the quoted string doesn't terminate and there's a backslash in there, but it doesn't terminate with a double quote.
Now this is obv bad, but trying to be lenient, and at least just giving the entire information back would be better than only giving the string after the backslash. (i.e. if I read this right, the following header example will only give "only this will be returned, I think" as the `filename`)

`Content-Disposition: form-data; name="yaml"; filename="this will be dropped, \!only this will be returned\r\n`
Please add a test that makes sure the filename from the above header results in:
`this will be dropped, !only this will be returned`